File: php/tests/GraphDocumentTest.php
<?php

declare(strict_types=1);

namespace Oojs\Tests;

use Oojs\ErrorCode;
use Oojs\Registry;
use Oojs\Validator;
use PHPUnit\Framework\TestCase;

/**
 * Tests for graph document validation (§8.12).
 *
 * A graph document serialises multiple OOJS objects into one JSON envelope:
 *   - "roots": entry-point objects (each with "$type" and "$id")
 *   - "objects": map of "$id" → additional objects referenced via { "$ref-id": "…" }
 *
 * Horizontal TypeRef properties hold { "$ref-id": "id" } instead of an embedded object.
 * The validator type-checks each reference without recursively validating the target
 * inline — targets are validated at the top level of the document.  This design
 * makes cycle handling (§8.13) implicit: no visited-set bookkeeping is required.
 */
class GraphDocumentTest extends TestCase
{
    private function employeeSchema(): array
    {
        return [
            '$oojs' => '1.0',
            '$id'   => 'https://example.org/schemas/employees',
            'types' => [
                'Department' => [
                    'properties' => [
                        'departmentId' => ['type' => 'string'],
                        'name'         => ['type' => 'string'],
                    ],
                    'required' => ['departmentId', 'name'],
                ],
                'Employee' => [
                    'properties' => [
                        'employeeId' => ['type' => 'string'],
                        'name'       => ['type' => 'string'],
                        'department' => ['type' => 'Department'],  // HAS-ONE → Department
                        'manager'    => ['type' => 'Employee'],    // optional HAS-ONE → Employee (self-ref)
                    ],
                    'required' => ['employeeId', 'name', 'department'],
                ],
            ],
        ];
    }

    // ------------------------------------------------------------------
    // Valid graph documents
    // ------------------------------------------------------------------

    public function test_two_employees_sharing_one_department(): void
    {
        // Alice and Bob both reference the same dept-eng object — stored once in "objects".
        $r      = new Registry();
        $schema = $r->loadDict($this->employeeSchema());

        $graphDoc = [
            '$oojs'   => '1.0',
            'roots'   => [
                [
                    '$type'      => 'Employee',
                    '$id'        => 'emp-alice',
                    'employeeId' => 'E-001',
                    'name'       => 'Alice',
                    'department' => ['$ref-id' => 'dept-eng'],
                ],
                [
                    '$type'      => 'Employee',
                    '$id'        => 'emp-bob',
                    'employeeId' => 'E-002',
                    'name'       => 'Bob',
                    'department' => ['$ref-id' => 'dept-eng'],
                ],
            ],
            'objects' => [
                'dept-eng' => [
                    '$type'        => 'Department',
                    '$id'          => 'dept-eng',
                    'departmentId' => 'D-01',
                    'name'         => 'Engineering',
                ],
            ],
        ];

        $v    = new Validator($r);
        $errs = $v->validateGraphDocument($graphDoc, $schema);
        $this->assertSame([], $errs);
    }

    public function test_graph_with_optional_manager_reference(): void
    {
        // Carol's manager is Alice, Alice has no manager.
        $r      = new Registry();
        $schema = $r->loadDict($this->employeeSchema());

        $graphDoc = [
            '$oojs'   => '1.0',
            'roots'   => [
                [
                    '$type'      => 'Employee',
                    '$id'        => 'emp-alice',
                    'employeeId' => 'E-001',
                    'name'       => 'Alice',
                    'department' => ['$ref-id' => 'dept-eng'],
                    // no 'manager' — optional field absent
                ],
                [
                    '$type'      => 'Employee',
                    '$id'        => 'emp-carol',
                    'employeeId' => 'E-003',
                    'name'       => 'Carol',
                    'department' => ['$ref-id' => 'dept-eng'],
                    'manager'    => ['$ref-id' => 'emp-alice'],
                ],
            ],
            'objects' => [
                'dept-eng' => [
                    '$type'        => 'Department',
                    '$id'          => 'dept-eng',
                    'departmentId' => 'D-01',
                    'name'         => 'Engineering',
                ],
            ],
        ];

        $v    = new Validator($r);
        $errs = $v->validateGraphDocument($graphDoc, $schema);
        $this->assertSame([], $errs);
    }

    public function test_graph_with_no_objects_map(): void
    {
        // Roots only, no "objects" key — each root has no $ref-id references.
        $r      = new Registry();
        $schema = $r->loadDict([
            '$oojs' => '1.0',
            '$id'   => 'https://example.org/schemas/simple',
            'types' => [
                'Tag' => [
                    'properties' => ['tagId' => ['type' => 'string'], 'label' => ['type' => 'string']],
                    'required'   => ['tagId', 'label'],
                ],
            ],
        ]);

        $graphDoc = [
            '$oojs' => '1.0',
            'roots' => [
                ['$type' => 'Tag', '$id' => 'tag-1', 'tagId' => 'T-1', 'label' => 'alpha'],
                ['$type' => 'Tag', '$id' => 'tag-2', 'tagId' => 'T-2', 'label' => 'beta'],
            ],
        ];

        $v    = new Validator($r);
        $errs = $v->validateGraphDocument($graphDoc, $schema);
        $this->assertSame([], $errs);
    }

    // ------------------------------------------------------------------
    // §8.13.2 — Cycles in graph documents
    // ------------------------------------------------------------------

    public function test_mutual_manager_cycle_does_not_hang(): void
    {
        // Alice's manager is Bob; Bob's manager is Alice — a two-node cycle.
        // Validation must terminate and report no errors (both are structurally valid).
        $r      = new Registry();
        $schema = $r->loadDict($this->employeeSchema());

        $graphDoc = [
            '$oojs'   => '1.0',
            'roots'   => [
                [
                    '$type'      => 'Employee',
                    '$id'        => 'emp-alice',
                    'employeeId' => 'E-001',
                    'name'       => 'Alice',
                    'department' => ['$ref-id' => 'dept-eng'],
                    'manager'    => ['$ref-id' => 'emp-bob'],
                ],
                [
                    '$type'      => 'Employee',
                    '$id'        => 'emp-bob',
                    'employeeId' => 'E-002',
                    'name'       => 'Bob',
                    'department' => ['$ref-id' => 'dept-eng'],
                    'manager'    => ['$ref-id' => 'emp-alice'],  // points back to Alice
                ],
            ],
            'objects' => [
                'dept-eng' => [
                    '$type'        => 'Department',
                    '$id'          => 'dept-eng',
                    'departmentId' => 'D-01',
                    'name'         => 'Engineering',
                ],
            ],
        ];

        $v    = new Validator($r);
        $errs = $v->validateGraphDocument($graphDoc, $schema);
        $this->assertSame([], $errs);
    }

    public function test_three_node_cycle_does_not_hang(): void
    {
        // A.manager → B, B.manager → C, C.manager → A.
        $r      = new Registry();
        $schema = $r->loadDict($this->employeeSchema());

        $graphDoc = [
            '$oojs'   => '1.0',
            'roots'   => [
                [
                    '$type'      => 'Employee',
                    '$id'        => 'emp-a',
                    'employeeId' => 'E-A',
                    'name'       => 'Alpha',
                    'department' => ['$ref-id' => 'dept-eng'],
                    'manager'    => ['$ref-id' => 'emp-b'],
                ],
                [
                    '$type'      => 'Employee',
                    '$id'        => 'emp-b',
                    'employeeId' => 'E-B',
                    'name'       => 'Beta',
                    'department' => ['$ref-id' => 'dept-eng'],
                    'manager'    => ['$ref-id' => 'emp-c'],
                ],
                [
                    '$type'      => 'Employee',
                    '$id'        => 'emp-c',
                    'employeeId' => 'E-C',
                    'name'       => 'Gamma',
                    'department' => ['$ref-id' => 'dept-eng'],
                    'manager'    => ['$ref-id' => 'emp-a'],  // closes the cycle
                ],
            ],
            'objects' => [
                'dept-eng' => [
                    '$type'        => 'Department',
                    '$id'          => 'dept-eng',
                    'departmentId' => 'D-01',
                    'name'         => 'Engineering',
                ],
            ],
        ];

        $v    = new Validator($r);
        $errs = $v->validateGraphDocument($graphDoc, $schema);
        $this->assertSame([], $errs);
    }

    // ------------------------------------------------------------------
    // Invalid graph documents
    // ------------------------------------------------------------------

    public function test_missing_required_field_in_root_object(): void
    {
        $r      = new Registry();
        $schema = $r->loadDict($this->employeeSchema());

        $graphDoc = [
            '$oojs'   => '1.0',
            'roots'   => [
                [
                    '$type'      => 'Employee',
                    '$id'        => 'emp-alice',
                    // 'employeeId' missing — required
                    'name'       => 'Alice',
                    'department' => ['$ref-id' => 'dept-eng'],
                ],
            ],
            'objects' => [
                'dept-eng' => [
                    '$type'        => 'Department',
                    '$id'          => 'dept-eng',
                    'departmentId' => 'D-01',
                    'name'         => 'Engineering',
                ],
            ],
        ];

        $v    = new Validator($r);
        $errs = $v->validateGraphDocument($graphDoc, $schema);
        $this->assertNotEmpty($errs);
        $codes = array_column($errs, 'code');
        $this->assertContains(ErrorCode::MISSING_REQUIRED, $codes);
    }

    public function test_missing_required_field_in_objects_map(): void
    {
        $r      = new Registry();
        $schema = $r->loadDict($this->employeeSchema());

        $graphDoc = [
            '$oojs'   => '1.0',
            'roots'   => [
                [
                    '$type'      => 'Employee',
                    '$id'        => 'emp-alice',
                    'employeeId' => 'E-001',
                    'name'       => 'Alice',
                    'department' => ['$ref-id' => 'dept-eng'],
                ],
            ],
            'objects' => [
                'dept-eng' => [
                    '$type' => 'Department',
                    '$id'   => 'dept-eng',
                    // 'departmentId' and 'name' both missing
                ],
            ],
        ];

        $v    = new Validator($r);
        $errs = $v->validateGraphDocument($graphDoc, $schema);
        $this->assertNotEmpty($errs);
        $codes = array_column($errs, 'code');
        $this->assertContains(ErrorCode::MISSING_REQUIRED, $codes);
    }

    public function test_unresolved_ref_id_produces_error(): void
    {
        $r      = new Registry();
        $schema = $r->loadDict($this->employeeSchema());

        $graphDoc = [
            '$oojs' => '1.0',
            'roots' => [
                [
                    '$type'      => 'Employee',
                    '$id'        => 'emp-alice',
                    'employeeId' => 'E-001',
                    'name'       => 'Alice',
                    'department' => ['$ref-id' => 'does-not-exist'],  // no such $id
                ],
            ],
        ];

        $v    = new Validator($r);
        $errs = $v->validateGraphDocument($graphDoc, $schema);
        $this->assertNotEmpty($errs);
        $codes = array_column($errs, 'code');
        $this->assertContains(ErrorCode::UNRESOLVED_REFERENCE, $codes);
    }

    public function test_ref_id_type_mismatch_produces_error(): void
    {
        // Employee.department expects a Department, but dept-eng has $type Employee.
        $r      = new Registry();
        $schema = $r->loadDict($this->employeeSchema());

        $graphDoc = [
            '$oojs'   => '1.0',
            'roots'   => [
                [
                    '$type'      => 'Employee',
                    '$id'        => 'emp-alice',
                    'employeeId' => 'E-001',
                    'name'       => 'Alice',
                    'department' => ['$ref-id' => 'not-a-dept'],
                ],
            ],
            'objects' => [
                'not-a-dept' => [
                    '$type'      => 'Employee',  // wrong type for a department slot
                    '$id'        => 'not-a-dept',
                    'employeeId' => 'E-999',
                    'name'       => 'Impostor',
                    'department' => ['$ref-id' => 'not-a-dept'],  // self-ref to keep it valid on its own
                ],
            ],
        ];

        $v    = new Validator($r);
        $errs = $v->validateGraphDocument($graphDoc, $schema);
        $this->assertNotEmpty($errs);
        $codes = array_column($errs, 'code');
        $this->assertContains(ErrorCode::TYPE_MISMATCH, $codes);
    }

    public function test_root_missing_type_produces_error(): void
    {
        $r      = new Registry();
        $schema = $r->loadDict($this->employeeSchema());

        $graphDoc = [
            '$oojs' => '1.0',
            'roots' => [
                [
                    // no '$type' field
                    '$id'        => 'emp-alice',
                    'employeeId' => 'E-001',
                    'name'       => 'Alice',
                ],
            ],
        ];

        $v    = new Validator($r);
        $errs = $v->validateGraphDocument($graphDoc, $schema);
        $this->assertNotEmpty($errs);
        $codes = array_column($errs, 'code');
        $this->assertContains(ErrorCode::MISSING_DISCRIMINATOR, $codes);
    }

    // ------------------------------------------------------------------
    // File-based integration: graph-document example
    // ------------------------------------------------------------------

    public function test_graph_document_example_file(): void
    {
        $examples = __DIR__ . '/../../examples';
        $r        = new Registry();
        $schema   = $r->loadFile("$examples/graph-document.oojs.json");
        $graphDoc = json_decode(
            file_get_contents("$examples/graph-document.json"),
            associative: true,
            flags: JSON_THROW_ON_ERROR,
        );

        // Strip top-level _comment (not a graph document field)
        unset($graphDoc['_comment']);

        $v    = new Validator($r);
        $errs = $v->validateGraphDocument($graphDoc, $schema);
        $this->assertSame([], $errs, 'graph-document.json should be valid: ' . implode(', ', array_map('strval', $errs)));
    }
}

File: java/src/test/java/org/oojs/GraphDocumentTest.java
package org.oojs;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.Test;
import org.oojs.model.Schema;

import java.io.File;
import java.nio.file.Paths;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

class GraphDocumentTest extends TestHelpers {

    private static final ObjectMapper MAPPER = new ObjectMapper();

    private Map<String, Object> employeeSchema() {
        return map(
                "$oojs", "1.0",
                "$id", "https://example.org/schemas/employees",
                "types", map(
                        "Department", map(
                                "properties", map(
                                        "departmentId", map("type", "string"),
                                        "name", map("type", "string")),
                                "required", list("departmentId", "name")),
                        "Employee", map(
                                "properties", map(
                                        "employeeId", map("type", "string"),
                                        "name", map("type", "string"),
                                        "department", map("type", "Department"),
                                        "manager", map("type", "Employee")),
                                "required", list("employeeId", "name", "department"))));
    }

    @Test
    void two_employees_sharing_one_department() {
        Registry r = new Registry();
        Schema schema = r.loadMap(employeeSchema());

        Map<String, Object> graphDoc = map(
                "$oojs", "1.0",
                "roots", list(
                        map("$type", "Employee", "$id", "emp-alice", "employeeId", "E-001", "name", "Alice",
                                "department", map("$ref-id", "dept-eng")),
                        map("$type", "Employee", "$id", "emp-bob", "employeeId", "E-002", "name", "Bob",
                                "department", map("$ref-id", "dept-eng"))),
                "objects", map(
                        "dept-eng", map("$type", "Department", "$id", "dept-eng", "departmentId", "D-01", "name", "Engineering")));

        List<ValidationError> errs = new Validator(r).validateGraphDocument(graphDoc, schema);
        assertEquals(List.of(), errs);
    }

    @Test
    void graph_with_optional_manager_reference() {
        Registry r = new Registry();
        Schema schema = r.loadMap(employeeSchema());

        Map<String, Object> graphDoc = map(
                "$oojs", "1.0",
                "roots", list(
                        map("$type", "Employee", "$id", "emp-alice", "employeeId", "E-001", "name", "Alice",
                                "department", map("$ref-id", "dept-eng")),
                        map("$type", "Employee", "$id", "emp-carol", "employeeId", "E-003", "name", "Carol",
                                "department", map("$ref-id", "dept-eng"), "manager", map("$ref-id", "emp-alice"))),
                "objects", map(
                        "dept-eng", map("$type", "Department", "$id", "dept-eng", "departmentId", "D-01", "name", "Engineering")));

        List<ValidationError> errs = new Validator(r).validateGraphDocument(graphDoc, schema);
        assertEquals(List.of(), errs);
    }

    @Test
    void graph_with_no_objects_map() {
        Registry r = new Registry();
        Schema schema = r.loadMap(map(
                "$oojs", "1.0",
                "$id", "https://example.org/schemas/simple",
                "types", map("Tag", map(
                        "properties", map("tagId", map("type", "string"), "label", map("type", "string")),
                        "required", list("tagId", "label")))));

        Map<String, Object> graphDoc = map(
                "$oojs", "1.0",
                "roots", list(
                        map("$type", "Tag", "$id", "tag-1", "tagId", "T-1", "label", "alpha"),
                        map("$type", "Tag", "$id", "tag-2", "tagId", "T-2", "label", "beta")));

        List<ValidationError> errs = new Validator(r).validateGraphDocument(graphDoc, schema);
        assertEquals(List.of(), errs);
    }

    @Test
    void mutual_manager_cycle_does_not_hang() {
        Registry r = new Registry();
        Schema schema = r.loadMap(employeeSchema());

        Map<String, Object> graphDoc = map(
                "$oojs", "1.0",
                "roots", list(
                        map("$type", "Employee", "$id", "emp-alice", "employeeId", "E-001", "name", "Alice",
                                "department", map("$ref-id", "dept-eng"), "manager", map("$ref-id", "emp-bob")),
                        map("$type", "Employee", "$id", "emp-bob", "employeeId", "E-002", "name", "Bob",
                                "department", map("$ref-id", "dept-eng"), "manager", map("$ref-id", "emp-alice"))),
                "objects", map(
                        "dept-eng", map("$type", "Department", "$id", "dept-eng", "departmentId", "D-01", "name", "Engineering")));

        List<ValidationError> errs = new Validator(r).validateGraphDocument(graphDoc, schema);
        assertEquals(List.of(), errs);
    }

    @Test
    void three_node_cycle_does_not_hang() {
        Registry r = new Registry();
        Schema schema = r.loadMap(employeeSchema());

        Map<String, Object> graphDoc = map(
                "$oojs", "1.0",
                "roots", list(
                        map("$type", "Employee", "$id", "emp-a", "employeeId", "E-A", "name", "Alpha",
                                "department", map("$ref-id", "dept-eng"), "manager", map("$ref-id", "emp-b")),
                        map("$type", "Employee", "$id", "emp-b", "employeeId", "E-B", "name", "Beta",
                                "department", map("$ref-id", "dept-eng"), "manager", map("$ref-id", "emp-c")),
                        map("$type", "Employee", "$id", "emp-c", "employeeId", "E-C", "name", "Gamma",
                                "department", map("$ref-id", "dept-eng"), "manager", map("$ref-id", "emp-a"))),
                "objects", map(
                        "dept-eng", map("$type", "Department", "$id", "dept-eng", "departmentId", "D-01", "name", "Engineering")));

        List<ValidationError> errs = new Validator(r).validateGraphDocument(graphDoc, schema);
        assertEquals(List.of(), errs);
    }

    @Test
    void missing_required_field_in_root_object() {
        Registry r = new Registry();
        Schema schema = r.loadMap(employeeSchema());

        Map<String, Object> graphDoc = map(
                "$oojs", "1.0",
                "roots", list(
                        map("$type", "Employee", "$id", "emp-alice", "name", "Alice",
                                "department", map("$ref-id", "dept-eng"))),
                "objects", map(
                        "dept-eng", map("$type", "Department", "$id", "dept-eng", "departmentId", "D-01", "name", "Engineering")));

        List<ValidationError> errs = new Validator(r).validateGraphDocument(graphDoc, schema);
        assertFalse(errs.isEmpty());
        assertTrue(hasCode(errs, ErrorCode.MISSING_REQUIRED));
    }

    @Test
    void missing_required_field_in_objects_map() {
        Registry r = new Registry();
        Schema schema = r.loadMap(employeeSchema());

        Map<String, Object> graphDoc = map(
                "$oojs", "1.0",
                "roots", list(
                        map("$type", "Employee", "$id", "emp-alice", "employeeId", "E-001", "name", "Alice",
                                "department", map("$ref-id", "dept-eng"))),
                "objects", map(
                        "dept-eng", map("$type", "Department", "$id", "dept-eng")));

        List<ValidationError> errs = new Validator(r).validateGraphDocument(graphDoc, schema);
        assertFalse(errs.isEmpty());
        assertTrue(hasCode(errs, ErrorCode.MISSING_REQUIRED));
    }

    @Test
    void unresolved_ref_id_produces_error() {
        Registry r = new Registry();
        Schema schema = r.loadMap(employeeSchema());

        Map<String, Object> graphDoc = map(
                "$oojs", "1.0",
                "roots", list(
                        map("$type", "Employee", "$id", "emp-alice", "employeeId", "E-001", "name", "Alice",
                                "department", map("$ref-id", "does-not-exist"))));

        List<ValidationError> errs = new Validator(r).validateGraphDocument(graphDoc, schema);
        assertFalse(errs.isEmpty());
        assertTrue(hasCode(errs, ErrorCode.UNRESOLVED_REFERENCE));
    }

    @Test
    void ref_id_type_mismatch_produces_error() {
        Registry r = new Registry();
        Schema schema = r.loadMap(employeeSchema());

        Map<String, Object> graphDoc = map(
                "$oojs", "1.0",
                "roots", list(
                        map("$type", "Employee", "$id", "emp-alice", "employeeId", "E-001", "name", "Alice",
                                "department", map("$ref-id", "not-a-dept"))),
                "objects", map(
                        "not-a-dept", map("$type", "Employee", "$id", "not-a-dept", "employeeId", "E-999", "name", "Impostor",
                                "department", map("$ref-id", "not-a-dept"))));

        List<ValidationError> errs = new Validator(r).validateGraphDocument(graphDoc, schema);
        assertFalse(errs.isEmpty());
        assertTrue(hasCode(errs, ErrorCode.TYPE_MISMATCH));
    }

    @Test
    void root_missing_type_produces_error() {
        Registry r = new Registry();
        Schema schema = r.loadMap(employeeSchema());

        Map<String, Object> graphDoc = map(
                "$oojs", "1.0",
                "roots", list(
                        map("$id", "emp-alice", "employeeId", "E-001", "name", "Alice")));

        List<ValidationError> errs = new Validator(r).validateGraphDocument(graphDoc, schema);
        assertFalse(errs.isEmpty());
        assertTrue(hasCode(errs, ErrorCode.MISSING_DISCRIMINATOR));
    }

    @Test
    @SuppressWarnings("unchecked")
    void graph_document_example_file() throws Exception {
        String examples = Paths.get(System.getProperty("user.dir")).getParent().resolve("examples").toString();
        Registry r = new Registry();
        Schema schema = r.loadFile(examples + "/graph-document.oojs.json");
        Map<String, Object> graphDoc = MAPPER.readValue(new File(examples + "/graph-document.json"), Map.class);

        Map<String, Object> cleanDoc = new LinkedHashMap<>(graphDoc);
        cleanDoc.remove("_comment");

        List<ValidationError> errs = new Validator(r).validateGraphDocument(cleanDoc, schema);
        assertEquals(List.of(), errs);
    }
}
